tls_codec/derive/src/lib.rs

```diff
@@ -477,17 +529,45 @@ fn define_discriminant_constants(
             })
             .collect::<Result<Vec<_>>>()?
     } else {
-        variants
-            .iter()
-            .try_fold((0, Vec::new()), |(next, mut acc), variant| {
-                let constant_id = discriminant_id(&variant.ident);
-                let value = discriminant_value(&variant.attrs)?.unwrap_or(next);
-                acc.push(quote! {
-                    const #constant_id: #repr = #value as #repr;
-                });
-                Ok::<_, syn::Error>((value + 1, acc))
-            })?
-            .1
+        let mut spans = Vec::with_capacity(variants.len());
+        let mut implicit_discriminant = 0u32;
+        let mut discriminant_has_paths = false;
+        for variant in variants.iter() {
+            let constant_id = discriminant_id(&variant.ident);
+
+            let tokens = if let Some(value) = discriminant_value(&variant.attrs)? {
+                match value {
+                    DiscriminantValue::Literal(value_u32) => {
+                        implicit_discriminant = value_u32;
+                        quote! {
+                            const #constant_id: #repr = #value_u32 as #repr;
+                        }
+                    }
+                    DiscriminantValue::Path(pathexpr) => {
+                        discriminant_has_paths = true;
+                        quote! {
+                            const #constant_id: #repr = #pathexpr as #repr;
```

Comment:
You are forcing the path value into the representation of the enum. But this will lead to unexpected behaviour.
Taking the test in `encode.rs` for example you you can change `Potato = 0x0004,` to `Potato = 0x1000_0004`, i.e. u32, which is forced into `u16`. So
```rust
    let serialized = EnumWithDataAndConstDiscriminant::B
        .tls_serialize_detached()
        .unwrap();
    assert_eq!(vec![0, 4], serialized);
```
still holds, while I'd expect a compiler error that tells me that `Potato` is `u32` but `u16` was expected.

Reply:
I could just remove the `as #repr` part out of `#pathexpr as #repr` so that the compiler would error out because types differ instead of silently casting integers for convenience. Would that sound fine to you?

Reply:
After checking with a u64::MAX casted on a repr(u8) enum, you're absolutely right, this is not only unexpected behaviour but also Undefined Behaviour as well. Good catch, that's 100% a bug.

Reply:
By the way, the same issue exists with explicit literal discriminants as they are converted to `u32` first and *then* coerced into the enum's `repr`.